packages/vite/src/node/server/pluginContainer.ts

```diff
@@ -239,202 +265,402 @@ export async function createEnvironmentPluginContainer(
     await Promise.all(parallelPromises)
   }
 
-  // throw when an unsupported ModuleInfo property is accessed,
-  // so that incompatible plugins fail in a non-cryptic way.
-  const ModuleInfoProxy: ProxyHandler<ModuleInfo> = {
-    get(info: any, key: string) {
-      if (key in info) {
-        return info[key]
-      }
-      // Don't throw an error when returning from an async function
-      if (key === 'then') {
-        return undefined
-      }
-      throw Error(
-        `[vite] The "${key}" property of ModuleInfo is not supported.`,
-      )
-    },
+  async buildStart(_options?: InputOptions): Promise<void> {
+    await this.handleHookPromise(
+      this.hookParallel(
+        'buildStart',
+        (plugin) => this._getPluginContext(plugin),
+        () => [this.options as NormalizedInputOptions],
+      ),
+    )
   }
 
-  // same default value of "moduleInfo.meta" as in Rollup
-  const EMPTY_OBJECT = Object.freeze({})
-
-  // we should create a new context for each async hook pipeline so that the
-  // active plugin in that pipeline can be tracked in a concurrency-safe manner.
-  // using a class to make creating new contexts more efficient
-  class Context implements PluginContext {
-    environment: PluginEnvironment // TODO: | ScanEnvironment
-    meta = minimalContext.meta
-    ssr = false
-    _scan = false
-    _activePlugin: Plugin | null
-    _activeId: string | null = null
-    _activeCode: string | null = null
-    _resolveSkips?: Set<Plugin>
-    _addedImports: Set<string> | null = null
-
-    constructor(initialPlugin?: Plugin) {
-      this.environment = environment
-      this._activePlugin = initialPlugin || null
-    }
+  async resolveId(
+    rawId: string,
+    importer: string | undefined = join(
+      this.environment.config.root,
+      'index.html',
+    ),
+    options?: {
+      attributes?: Record<string, string>
+      custom?: CustomPluginOptions
+      skip?: Set<Plugin>
+      /**
+       * @internal
+       */
+      scan?: boolean
+      isEntry?: boolean
+    },
+  ): Promise<PartialResolvedId | null> {
+    const skip = options?.skip
+    const scan = !!options?.scan
+    const ssr = this.environment.name !== 'client'
+
+    const resolveStart = debugResolve ? performance.now() : 0
+    let id: string | null = null
+    const partial: Partial<PartialResolvedId> = {}
+    for (const plugin of this._utils.getSortedPlugins('resolveId')) {
+      if (this._closed && this.environment?.options.dev.recoverable)
+        throwClosedServerError()
+      if (!plugin.resolveId) continue
+      if (skip?.has(plugin)) continue
+
+      const ctx = this._getPluginContext(plugin)
+      ctx._resolveSkips = skip
+      ctx._scan = scan
 
```

Comment:
Re `_resolveSkips`, is this safe? Wouldn't it leak and share to other hooks too, and then `this.resolve()` would incorrectly skip the plugin?

Reply:
You are right, I fixed it in #17288 by having per-hook context. Rollup does this as well.